### ISSUE TITLE
[2.6.2] Regression : rights on forms after ticket creation

### ISSUE THREAD

**Issue by nqb** [pluginsGLPI/formcreator#904]
Hello,

After an upgrade from GLPI 9.1.6, formcreator 2.5.2 to GLPI 9.2.1, formcreator 2.6.1, rights on forms not working as expected.

Steps to reproduce : 
* Create a ticket via a form as Foo user (self-service profile)

Expected result : 
* As technician user (admin profile), you can see a link between form and ticket on ticket tab.
If you click on form, you see filled form.

Actual results : 
* As technician user (admin profile) on ticket :  link in Ticket tab is not "clickable"
* As technician user (admin profile) in Administration -> Forms -> My Form -> Form answers : there is no new form answer.
* As Foo user on ticket : link in Ticket tab is "clickable"
* As super-admin user in Administration -> Forms -> My Form -> Form answers : you see new form answer. Howewer, after a click on it, you get "You don't have permission to perform this action."

**Comment by nqb** [comment]
Same issue in 2.6.2. No changes on actual results.

**Comment by nqb** [comment]
@btry : do you reproduce ?

**Comment by btry** [comment]
Hi

I did not investigated yet.

**Comment by mgallais76** [comment]
Hello,

I have the exact same problem, do you fix it in the 2.6.3 version?

**Comment by btry** [comment]
I did not had time to work in this yet. As you are now 2 to report the problem, then I mark it as bug.

This is a community issue, then it will be solved at best effort. To have quicker fixes, you may consider requesting support : contact@teclib.com

**Comment by btry** [comment]
By the way, @nqb @mgallais76 did you tried  the release 2.6.2 ?

**Comment by nqb** [comment]
@btry : see my https://github.com/pluginsGLPI/formcreator/issues/904#issuecomment-364932301

**Comment by mgallais76** [comment]
Yes, I'm in the release 2.6.2.

**Comment by btry** [comment]
OK, thanks to both. I'll try to reproduce with the development version.

**Comment by btry** [comment]
- I'm using the admin profile (configured as a fresh GLPI instance).
- I added the UPDATE right on the profile to access the administrative part of Formcreator
- I added a user  **admin** to this profile
- I created a form (empty) and answered it with the user **normal**
- I login with **admin**, open the form for the test, and then  the tab **Form answers**

## Actual results :

### As technician user (admin profile) on ticket : link in Ticket tab is not "clickable"
I reproduce

### As technician user (admin profile) in Administration -> Forms -> My Form -> Form answers : there is no new form answer.
I don't reproduce (maybe an entity constraint between the user, the ticket and form answers set ?
![image](https://user-images.githubusercontent.com/14139801/38813783-3175e18e-4190-11e8-9126-abf31b511a1c.png)

### As Foo user on ticket : link in Ticket tab is "clickable"
I reproduce

### As super-admin user in Administration -> Forms -> My Form -> Form answers : you see new form answer. Howewer, after a click on it, you get "You don't have permission to perform this action."

I reproduce

**Comment by btry** [comment]
This regression was implemented as a feature. It solves confidentiality issues, and it might be too much restrictive.

Currently, only the requester or a validator (as is or member of a validator group) can see the form answers set.

I can allow the user to view the answers set if he can view the generated ticket or change. But if the answers set is used to generate several  tickets or changes, each of the containing a subset of the answers, there may be confidentiality issues.  

Example: a form is designed to generate a ticket for the technician A, and a ticket to the technician B. Each ticket have half of the fields of the answer set, and the tickets are generated different entities. One entity is for technician A and an other for t echnician B. These entities prevents the technicians to see the ticket for the other technician, for confidentiality reason.

If I allow to see the answers set, then I disclose the content both tickets to both technicians.

However, I need to find the best permissions for the administrators, probably by checking the rights to update an entity and the right to read tickets.

This needs to be analyzed with respect with current use of the plugin. FYI this change was developed 4 months ago.

**Comment by btry** [comment]
I reference this comment : https://github.com/pluginsGLPI/formcreator/issues/869#issuecomment-354332559

**Comment by btry** [comment]
I understand  that an admin or a super admin may need to access the answers set. I'll search how to properly implement that. I think I'll need to create a new right.

For technician profile I don't understand why it is necessary to access the answers set. A technician already has all the needed information in the ticket he can read.  Please describe your use case.

**Comment by nqb** [comment]
Hello @btry,

To begin, thanks for you precise feedbacks. I understand problematic cases.

I try to reproduce again issue with GLPI 9.2.2 and Formcreator 2.6.2 :

### As technician user (admin profile) in Administration -> Forms -> My Form -> Form answers : there is no new form answer.

=> Finally, I see form answers. Howewer, after a click on it, you get "You don't have permission to perform this action." (identical to super-admin user)

Entity configuration : 

```
               Organisation name (parent entity)
                    |
                    |
-------------------------------------------
|                   |                     |
|                   |                     |
Entity A (ex : IT)  Entity B           Entity C
```
All forms create in parent entity with child-entities set to yes. They create tickets in child-entities (A, B and C).

In my last test, my admin user has Admin profile in entity A and try to see form answers that create tickets in entity A.

> For technician profile I don't understand why it is necessary to access the answers set. A technician already has all the needed information in the ticket he can read. Please describe your use case.

I'm not a long-time user of Formcreator, maybe @mgallais76 can reply to this answer better than me.
However, I see one use case : you don't copy all form information in ticket to avoid overloaded description, technicians read extra informations directly in answers set.

**Comment by btry** [comment]
> I see form answers. Howewer, after a click on it, you get "You don't have permission to perform this action." (identical to super-admin user)

This behavior matches the current implementation. 

> I see one use case : you don't copy all form information in ticket to avoid overloaded description, technicians read extra informations directly in answers set.

I think you should put in the ticket all information needed by the technician. If the ticket description is really long, you may customize your target ticket to organize better your fields. Instead of ##FULLFORM## build your own template and user the questions and answers tags. See pictures below

![image](https://user-images.githubusercontent.com/14139801/38857367-5ecee27c-4229-11e8-8a47-56f5ca876d16.png)
![image](https://user-images.githubusercontent.com/14139801/38857432-8b9ae878-4229-11e8-812a-d4ce274dd332.png)

**Comment by Sophal-IT** [comment]
Dear @btry 
If I update Form Version 2.6.2 to 2.6.3, it will affect to Form data or not. I have experience to upgrade it before, and I have lose all data of Form.
Wait your feedback.
Thanks
Sophal

**Comment by btry** [comment]
Hi

After continuing development of the plugin I found that it was too complex to implement upgrade in the same design as most plugins. The upgrade code was error prone, and I had to reorganize it. Unfortunately, the legacy upgrade system must be kept, in case of someone wants to upgrade from very old versions to 2.6.0 or later.

The transition was by itself error prone, but upgrading from versions 2.5.x to 2.6.0 or later should not cause much trouble.

Anyway I would always recommend to test an upgrade on a testing instance of your GLPI with a copy of the production database. If you find an issue in the upgrade process your feedback is welcome.

In your particular case, you upgrade from 2.6.2 to 2.6.3. I try as much as possible to not change the database design between versions x.y.z where only z changes. Upgrading from 2.6.2 to 2.6.3 is this scenario.

You may appreciate yourself the risk of the upgrade by reading this file: https://github.com/pluginsGLPI/formcreator/blob/develop/install/update_2.6.2_2.6.3.php

Most of the changes in the DB targets a simplification of the plugin when it must process characters with accents or non latin (Greek, Russian, Chinese, etc). So you should not have any major issue.

**Comment by btry** [comment]
Hi

I implemented a check a few months ago. Sory I did not updated this issue.

First, a user found a very old upgrade bug recently. If you want to upgrade from version < 2.6.0, consider adding the patch #1059 to the release 2.6.4 and feedback in issue #997.

Next, try to acess a form answer with a super-admin acount.

If you have skills with PHP, have a look into the method PluginFormcreatorForm_Answer::canViewItem()
https://github.com/pluginsGLPI/formcreator/blob/v2.6.4/inc/form_answer.class.php#L67

You will find details about restrictions. I would appreciate feedbacks to ensure I don't forget any use case.
